### PR TITLE
perf(storage): make get_key use vnode bitmap filter

### PR DESCRIPTION
**Pull request by soundOfDestiny**
## What's changed and what's your intention?

make get_key use vnode bitmap filter

### CODE REVIEW

File: src/storage/src/hummock/utils.rs [hunk 1]
```diff
@@ -112,3 +112,27 @@ where
...
+    full_key: &[u8],
+    value_meta: Option<u8>,
+    vnode_bitmap: &Vec<VNodeBitmap>,
+) -> bool {
+    if vnode_bitmap.is_empty() {
+        return false;
+    }
+    if let Some(vnode) = value_meta {
+        if let Some(table_id) = get_table_id(full_key) {
+            if let Ok(pos) = vnode_bitmap.binary_search_by_key(&table_id, |bitmap| bitmap.table_id)
```
Comment (BugenZhao) [inline src/storage/src/hummock/utils.rs:126]: We may use let chains here.

```rust
if let Some(b) = a && Some(c) = b {
```

File: src/storage/src/hummock/state_store.rs [hunk 2]
```diff
@@ -194,74 +194,83 @@ impl HummockStorage {
...
-}
-
-impl StateStore for HummockStorage {
-    type Iter = HummockStateStoreIter;
-
-    define_state_store_associated_type!();
 
     /// Gets the value of a specified `key`.
     /// The result is based on a snapshot corresponding to the given `epoch`.
     ///
```
Comment (BugenZhao) [inline src/storage/src/hummock/state_store.rs:200]: Could you also please add the doc for `value_meta` here?

File: src/storage/src/hummock/sstable/builder.rs [hunk 3]
```diff
@@ -115,11 +115,12 @@ impl SSTableBuilder {
...
-        let table_id = get_table_id(full_key);
-        self.vnode_bitmaps
-            .entry(table_id)
-            .or_insert([0; VNODE_BITMAP_LEN])[(value_meta >> 3) as usize] |=
-            1 << (value_meta & 0b111);
+        if let Some(table_id) = get_table_id(full_key) {
+            self.vnode_bitmaps
+                .entry(table_id)
+                .or_insert([0; VNODE_BITMAP_LEN])[(value_meta >> 3) as usize] |=
+                1 << (value_meta & 0b111);
```
Comment (BugenZhao) [inline src/storage/src/hummock/sstable/builder.rs:122]: Could you please add some comments for this?

File: src/storage/src/hummock/utils.rs (hunk 1 above)
Comment (soundOfDestiny) [reply src/storage/src/hummock/utils.rs:126]: thanks

File: src/storage/src/hummock/state_store.rs (hunk 2 above)
Comment (soundOfDestiny) [reply src/storage/src/hummock/state_store.rs:200]: fixed

File: src/storage/src/hummock/sstable/builder.rs (hunk 3 above)
Comment (soundOfDestiny) [reply src/storage/src/hummock/sstable/builder.rs:122]: fixed

File: src/storage/src/hummock/state_store.rs [hunk 4]
```diff
@@ -194,74 +194,84 @@ impl HummockStorage {
...
-                        for table_info in table_infos.into_iter().rev() {
+        for level in pinned_version.levels() {
+            if level.table_infos.is_empty() {
+                continue;
+            }
+            match level.level_type() {
+                LevelType::Overlapping => {
+                    let table_infos = prune_ssts(level.table_infos.iter(), &(key..=key), false);
+                    for table_info in table_infos.into_iter().rev() {
+                        if !vnode_bitmap_filter(
```
Comment (hzxa21) [inline src/storage/src/hummock/state_store.rs:254]: Using vnode bitmap to filter SSTs is a sst pruning strategy. Can we put this check in `prune_ssts` and only return ssts that pass the check?

Comment (soundOfDestiny) [reply src/storage/src/hummock/state_store.rs:254]: > Using vnode bitmap to filter SSTs is a sst pruning strategy. Can we put this check in `prune_ssts` and only return ssts that pass the check?

but parameter of `prune_ssts` is range of key, which could include lots of different vnode hashval

Comment (hzxa21) [reply src/storage/src/hummock/state_store.rs:254]: True. `prune_ssts` should filter SSTs by a set for vnodes. I think executor will also pass a set of vnodes it is interested at to scan/iter interfaces so we can use the same `prune_ssts` to do SST pruning.

Comment (soundOfDestiny) [reply src/storage/src/hummock/state_store.rs]: > True. `prune_ssts` should filter SSTs by a set for vnodes. I think executor will also pass a set of vnodes it is interested at to scan/iter interfaces so we can use the same `prune_ssts` to do SST pruning.

done

**Review by hzxa21 — COMMENTED** [review]
Please also add some unit tests for `get_with_vnode_set`

**Comment by soundOfDestiny** [comment]
> Please also add some unit tests for `get_with_vnode_set` to test get with Some(VNodeBitmap)

added

**Review by hzxa21 — APPROVED** [review]

**Pull request merged**